polkadot/xcm/xcm-builder/src/weight.rs

```diff
@@ -95,20 +95,20 @@ where
 	M: Get<u32>,
 	Instruction<C>: xcm::latest::GetWeight<W>,
 {
-	fn weight_with_limit(message: &Xcm<C>, instrs_limit: &mut u32) -> Result<Weight, ()> {
+	fn weight_with_limit(message: &mut Xcm<C>, instrs_limit: &mut u32) -> Result<Weight, ()> {
 		let mut r: Weight = Weight::zero();
 		*instrs_limit = instrs_limit.checked_sub(message.0.len() as u32).ok_or(())?;
-		for m in message.0.iter() {
-			r = r.checked_add(&Self::instr_weight_with_limit(m, instrs_limit)?).ok_or(())?;
+		for instruction in message.0.iter_mut() {
+			r = r.checked_add(&Self::instr_weight_with_limit(instruction, instrs_limit)?).ok_or(())?;
 		}
 		Ok(r)
 	}
 	fn instr_weight_with_limit(
-		instruction: &Instruction<C>,
+		instruction: &mut Instruction<C>,
 		instrs_limit: &mut u32,
 	) -> Result<Weight, ()> {
 		let instr_weight = match instruction {
-			Transact { require_weight_at_most, .. } => *require_weight_at_most,
+			Transact { ref mut call, .. } => call.take_decoded()?.get_dispatch_info().call_weight,
```

Comment:
```suggestion
			Transact { ref mut call, .. } => call.ensure_decoded()?.get_dispatch_info().call_weight,
```

I think `ensure_decoded` is better here because when you use `take_decoded` and the inner `DoubleEncoded.decoded` value is `None`, you trigger `T::decode_all_with_depth_limit` for the first time, but the inner `DoubleEncoded.decoded` remains `None`. 

Then, when it reaches the executor, we call `take_decoded` [here](https://github.com/paritytech/polkadot-sdk/pull/6228/files#diff-89914efafe7e34b9dc7fba0a2604f2c53a39e06e2fdc78e50376e8e5b19a892fR940-R941), but the inner value is still `None`, so we end up triggering `T::decode_all_with_depth_limit` again unnecessarily

Hopefully, I’m not missing anything, but we don’t use `ensure_decoded` (except in `DoubleEncoded::try_into`, which isn't used anywhere). So, iiuc, there’s no other way to mutate `DoubleEncoded.encoded/decoded`, so it should be fine and work as follows:
- `instr_weight_with_limit` -> `ensure_decoded`: ensures we only trigger decoding once and initializes the inner value.
- `execute(Transact)` -> `take_decoded`: retrieves the previously decoded value (if, for any reason, it wasn’t decoded, we can decode in-place as a fallback)."